Pokergameapp step5

Settpark opened the pull request:
#### 미션 진행 시 어려 웠던 점
* 게임의 방식, 인원 수를 고를 때 까지 함수 실행을 while 대기하고 싶었으나, 앱 자체가 실행이 안되어, 버튼을 선택할 때 마다 카드를 그리는 방식으로 대체하였습니다.
* 여태까지 진행한 카드 model의 Description("카드 문자열")을 view에서 받았는 데, 카드의 배열로 받아 져서 카드 한개 단위로 쪼개야 했었습니다.

#### 알게된 점
* main.storyboard에서 추가하던 Object들의 기능을 IBACTION뿐 아닌 objc로도 연결할 수 있다는 것이었습니다.

#### 개선하고 싶었던 점
* 사용자가 control 버튼 두가지 모두 누를 때까지 대기하고 함수 실행하지 않는 것이었습니다.
* segmentcontrol 부분의 함수 실행 부를 switch를 사용하지 않고 해결하는 점이었습니다.

Settpark committed:
카드 이미지 추가

Settpark committed:
이미지 추가전 수정한 사항 rebase 때문에 초기화 되어서 다시 추가

Settpark committed:
안쓰는 스위프트 파일 제거

안 쓰는 playManager 파일 제거

Settpark committed:
버튼 추가 후 버튼 클릭시 카드 이미지 등장하도록 수정

Settpark committed:
카드 위치 조정 및 카드 생성 방식 변경

players 에서 uiimageview를 만들어 반환하던 것을 카드의 이름만 반환하여 mainView에 넘기고, mainview에서 uiimageview를 생성, 후 카드가 계속 누적하여 새로 그려져서 뷰를 하나 만들고 해당 뷰에 카드만 그린 후, 다시 그릴 때 해당 뷰의 서브뷰를 전부 삭제하고, 다시 그리도록 함

Settpark committed:
readme.md 이미지 파일 추가

Settpark committed:
shake 버튼 추가

shake 버튼 추가하고 버튼 클릭시 카드 다시 섞도록 함

godrm commented on PokerGameApp/PokerGameApp/Card.swift:
```diff
@@ -17,7 +17,8 @@ enum Suit : String, CustomStringConvertible, CaseIterable {
     }
-    case spade = "♠︎", club = "♣︎", heart = "♥︎", diamond = "♦︎"
+    //case spade = "♠︎", club = "♣︎", heart = "♥︎", diamond = "♦︎"
+    case spade = "s", club = "c", heart = "h", diamond = "d"
```
아마도 카드 이미지 파일 이름 때문에 여기를 바꾼 것 같은데
이미지 파일에 대한 것은 자체는 Card 타입이 갖고 있어야 하는 속성이 아닙니다. 
하위 타입을 상위 타입에서 필요한 모든 것을 갖고 있어야 하는 게 아닙니다. 이 부분을 잘 생각해보세요.

Settpark replied:
player가 각자 자신의 카드를 직접그리도록 변경하였고, 그에 따라 해당 문제를 class Card -> class Player로 변경하였습니다.

godrm commented on PokerGameApp/PokerGameApp/Players.swift:
```diff
@@ -0,0 +1,43 @@
+    init(playerCount : Int){
+        self.players = []
+        self.playerCount = playerCount
+        print("플레이어 카운트 \(self.playerCount)")
```
개발 과정에서 디버깅이 필요하면 디버거 + 브레이크포인터를 활용하세요. 하위 모듈에서는 print를 직접 호출하지 않는 게 좋습니다.

godrm commented on PokerGameApp/PokerGameApp/Players.swift:
```diff
@@ -0,0 +1,43 @@
+    }
+    
+    func printCardplayers(index : Int) -> Void{
+        players[index].printMyCard()
```
중간 단계에 있는 타입은 출력 로직을 포함하지 않고 데이터를 전달해주는 방식을 제공하는 게 좋습니다. 
print보다 다른 방식으로 처리하는 경우를 고려해야 합니다.

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -8,40 +8,117 @@
+    }
+    
+    func drawCardinView(){
+        self.mainCardView.subviews.forEach{$0.removeFromSuperview()}
```
아무리 한 줄로 표현이 가능하더라도 `{ }` 스코프가 생기면 아래로 내려쓰는 게 좋습니다.

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -8,40 +8,117 @@
     
     func initGameplay(){
         let pattern : UIImage = UIImage(named: "bg_pattern")!
         self.view.backgroundColor = UIColor(patternImage: pattern)
```
옵셔널 강제 언래핑!은 사용하지 않는 게 좋습니다. 옵셔널 바인딩으로 개선해보세요.  
그렇지 않으면 파일 이름이 잘못된 경우 앱이 그냥 죽어버립니다.

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -8,40 +8,117 @@
+            break;
+        default:
+            break;
         }
```
이 부분의 switch-case를 제거할 수 있는 방법이 있습니다. 
gameStyle에 대한 타입을 선언해보세요. 그러면 GameType.init(segmentControl.selectedSegmentIndex) 형태로 줄일 수 있습니다.

Settpark replied:
주신 힌트로 해결하였습니다. 올바르게 작동하는 걸 확인할 수 있었습니다.

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -8,40 +8,117 @@
+    }
+    
+    func initGamestyleSegcon(){ // 게임 스타일 세그먼트 컨트롤 초기화
+        let gameStyle = ["7 Cards", "5 Cards"]
```
GameType를 선언하면 여기 `["7 Cards", "5 Cards"]` 값을 하드코딩 하지 않고 생성할 수 있는 방법이 있습니다.
Card.Rank 나 Card.Suit 를 참고해서 만들어보세요.

Settpark replied:
gamestyle.stud들 gamestyle.swift에 선언하여 만들고 해결하였습니다.

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -8,40 +8,117 @@
+        let CARDHEIGHT : CGFloat = CARDWIDTH * 1.27
+        game.gameStart()
+        for i in 0..<(self.players){
+            for j in 0..<(self.gameStyle){
```
여기 반복문, 배열을 탐색하는 것도 최상위 모듈에 해당하는 ViewController가 하지말고
game 이나 더 하위 모듈에서 처리하도록 하세요.
여기서는 Card별로 UIImage를 생성하기 위한 로직을 클로저 형태로 처리해보는 게 좋습니다.

Settpark replied:
하위 모듈에서 처리하도록 하였습니다. 다만, gamestyle에 따라, players에 따라 카드가 배치되는 방식이 바뀌기 때문에 pokergame에서 너무 많은 매개변수와, (뷰 컨트롤러) -> (포커게임) -> (게임스타일) -> (플레이어즈) -> (플레이어) -> (딜러) 이런식으로 연쇄적으로 메시지를 너무 많이 보내는 현상이 일어났습니다. 하위에서 연산해서 반환해서 주면 참 좋은데 단순히 플레이어 수가 궁금하고, 현재의 게임 스타일이 궁금한것이어서 그냥 하위로 보내서 처리하도록 하였습니다. 최상위 모듈에서는 다소 적은 매개변수를 사용하여서 조금은 위안이 되는데, 이 보다 더 나은 방법이 떠오르지 않아 이렇게 계속 메서드를 호출하는 것이 권장하지 않는 방법인지도 사실 잘 모르겠습니다. 만약 권장하지 않는 방법이라면 이를 해결하기 위한 방법이 있을까요?

godrm replied:
계층적으로 하위 모듈/객체를 호출하는게 당연한 겁니다. 
다만 동일한 형태 메소드를 계속해서 호출하는 건 그만큼 계층화가 덜 되서 일 수 있습니다.
하나의 타입이 바로 하위 타입 하나만 알도록 구조를 만드는 것은 상당수 괜찮습니다.

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -8,40 +8,117 @@
+        let OFFSETCENTER : Int = gameStyle/2 + 1
+        let OFFSET : CGFloat = CGFloat(self.view.frame.size.width/2) - CGFloat(Int(self.view.frame.width)/gameStyle * OFFSETCENTER) + CGFloat((Int(self.view.frame.width)/(self.gameStyle+1))/2)
+        let CARDWIDTH : CGFloat = self.view.frame.size.width/(CGFloat(gameStyle) + 1)
+        let CARDHEIGHT : CGFloat = CARDWIDTH * 1.27
```
CGFloat 타입이 엄청 보이네요. CGFloat 와 Int, Double도 계산을 자주 할 때 쓰는 것과 최종적으로 사용하는 경우에 따라서 타입을 구분해보세요. 모두가 CGFloat여야 하는건 아닐겁니다.

Settpark replied:
변경 해보려고 노력해 보았는데, 지금 보다 더 나은 기준 점이 생각나지 않아서 수정하지 못했습니다. CGRect에 들어가는 4요소를 CGFloat으로 기준을 잡다보니 (기준으로 잡은 이유는 self.view.frame.width 등이 CGFloat 형이기 때문에) 이런식으로 선언하게 된 것 같습니다. 혹시 구체적인 기준을 조금 알 수 있을까요?

godrm replied:
중간에 계산 과정에서 CGFloat을 쓰기보다는 다른 타입을 쓰다가 마지막에 할당할 때만 넣는 게 좋습니다.

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -8,40 +8,117 @@
+    
+    func drawCardinView(){
+        self.mainCardView.subviews.forEach{$0.removeFromSuperview()}
+        let game = PokerGame.init(GameStyle: self.gameStyle, gamePlayers: self.players) //게임 스타일 결정
```
근데 이렇게 구현하면 drawCardinView를 호출할 때마다 game이 새로 생기는건가요?
게임은 한 번만 만들고, CardDeck 카드가 나눠줄 수 있는 동안은 계속 게임 진행을 반복할 수 있어야 할텐데요?

Settpark replied:
PokerGame.swift의 CardShake함수를 통하여 수정하였습니다.

godrm reviewed (changes requested):
수고하셨습니다. 다음 피드백을 확인하고 개선해보세요.

Settpark committed:
pokergame 카드 분배와 rank, suit 변경

GameStyle 타입 생성, 그로 인해 pokergame에서 카드 나눠주는 방식 변경, rank와 suit -> card.rank, card.suit로 변경, TestCode에서 사용하던 takeRandomcard 함수 동작 방식 변경

Settpark committed:
포커 게임 카드 그리는 방식 수정 및 SegmentControl버튼에 따라 옵션 변경, Shake 버튼 기능 동작

각자의 플레이어가 카드 각자 직접 그리도록 수정, segment 스위치문 삭제를 위해 GameStyle 타입 선언 -> 코드 대부분 수정, seg로 선택 후 Shake버튼으로 카드 재분배하도록 설정함

Settpark committed:
불필요한 주석 제거, 게임 동작 방식 변경

시작 하자마자 카드를 나눠주는 게임 동작방식에서, Shake버튼을 누르면 나눠주도록 변경함, 처음에 초기화한 옵션을 볼 수 있도록 seg 옵션을 선택한 채로 시작하도록 변경함

Settpark committed:
pokergame의 불필요한 print문 제거

Settpark committed:
불필요한 print 함수 제거

Settpark committed:
foreach 스코프 수정

godrm commented on PokerGameApp/PokerGameApp/ViewController.swift:
```diff
@@ -8,70 +8,63 @@
-            drawCardinView()
-        default:
-            break;
+        players = Players.GamePlayers.allCases[segmentControl.selectedSegmentIndex]
```
이건 안전한 방법은 아니구요.
Players.GamePlayers.init(rawValue: segmentControl.selectedSegmentIndex) 형태가 더 안전합니다. 
참고하세요.

Settpark commented:
휴일에도 피드백 해주셔셔 감사합니다
